python/ray/data/_internal/memory_tracing.py

```diff
@@ -0,0 +1,140 @@
+"""Utility for debugging object store memory eager deletion in Datasets.
+
+Enable with RAY_DATASET_TRACE_ALLOCATIONS=1.
+
+Basic usage is to call `trace_allocation` each time a new object is created, and call
+`trace_deallocation` when an object should be disposed of. When the workload is
+complete, call `leak_report` to view possibly leaked objects.
+
+Note that so called "leaked" objects will be reclaimed eventually by reference counting
+in Ray. This is just to debug the eager deletion protocol which is more efficient.
+"""
+
+from io import StringIO
+from typing import Dict
+
+import ray
+from ray.data.context import DatasetContext
+
+
+def trace_allocation(ref: ray.ObjectRef, loc: str) -> None:
+    """Record that an object has been created.
+
+    Args:
+        ref: The object created.
+        loc: A human-readable string identifying the call site.
+    """
+    ctx = DatasetContext.get_current()
+    if ctx.trace_allocations:
+        tracer = _get_mem_actor()
+        ray.get(tracer.trace_alloc.remote([ref], loc))
+
+
+def trace_deallocation(ref: ray.ObjectRef, loc: str, free: bool = True) -> None:
+    """Record that an object has been deleted (and delete if if free=True).
+
+    Args:
+        ref: The object we no longer need.
+        loc: A human-readable string identifying the call site.
+        free: Whether to eagerly destroy the object instead of waiting for Ray
+            reference counting to kick in.
+    """
+    if free:
+        ray._private.internal_api.free(ref, local_only=False)
+    ctx = DatasetContext.get_current()
+    if ctx.trace_allocations:
+        tracer = _get_mem_actor()
+        ray.get(tracer.trace_dealloc.remote([ref], loc, free))
+
+
+def leak_report() -> str:
+    tracer = _get_mem_actor()
+    return ray.get(tracer.leak_report.remote())
+
+
+@ray.remote
+class _MemActor:
```

Comment:
is this memory actor something we plan to enable by default in the long term? Or just disable by default and enable when debugging is needed?

Wondering if the logic can be done in Dataset driver `DatasetStats` itself, or we must need another actor for it. All the logic looks pretty lightweight except fetching the object if size is missing.

As we are preparing to kill stats actor, I want to make sure we think twice before introducing another actor for Data.

Reply:
I don't think we'd ever enable this by default, since the performance overhead is fairly substantial. Or if we did, it would require more optimizations.

If you take a look at the full PR, I have added lighter weight memory stats per-operator to the graph, which can also be used for debugging. Though, it is not as comprehensive as this one, which covers both the new code and the legacy code.

Reply:
SG. shall we add comment in this file, to capture the caveat:

```
NOTE: the performance overhead of tracing object allocation is fairly substantial. This is meant to use in unit test for debugging. Please do not enable in production, without performance optimization.
```

Reply:
Added to main PR.